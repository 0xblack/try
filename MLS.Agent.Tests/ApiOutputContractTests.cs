using System.Threading.Tasks;
using Assent;
using Recipes;
using WorkspaceServer.Models;
using WorkspaceServer.Models.Execution;
using WorkspaceServer.Tests;
using Xunit;
using Xunit.Abstractions;

namespace MLS.Agent.Tests
{
    public class ApiOutputContractTests : ApiViaHttpTestsBase
    {
        private readonly Configuration configuration;

        public ApiOutputContractTests(ITestOutputHelper output) : base(output)
        {
            configuration = new Configuration()
                .UsingExtension("json");

#if !DEBUG
            configuration = configuration.SetInteractive(false);
#endif
        }

        [Fact]
        public async Task The_Run_contract_for_compiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode();

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var response = await CallRun(requestJson.ToJson());

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Run_contract_for_noncompiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode("doesn't compile");

            var request = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var requestBody = request.ToJson();

            var response = await CallRun(requestBody);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Completions_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Ou$$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallCompletion(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_signature_help_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Write($$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallSignatureHelp(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_instrumentation_contract_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: true)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_run_contract_with_no_instrumentation_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: false)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        private static Workspace.Buffer EntrypointCode(string mainContent = @"Console.WriteLine(Sample.Method());$$")
        {
            var input = $@"
using System;
using System.Linq;

namespace Example
{{
    public class Program
    {{
        public static void Main()
        {{
            {mainContent}
        }}       
    }}
}}".EnforceLF();
                
            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "EntrypointCode.cs",
                output,
                position ?? 0);
        }

        private static Workspace.Buffer ViewportCode(string methodContent = @"return ""Hello world!"";$$ ")
        {
            var input = $@"
using System.Collections.Generic;
using System;

namespace Example
{{
    public static class Sample
    {{
        public static object Method()
        {{
#region viewport
            {methodContent}
#endregion
        }}
    }}
}}".EnforceLF();

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "ViewportCode.cs",
                output,
                position ?? 0);
        }
    }
}