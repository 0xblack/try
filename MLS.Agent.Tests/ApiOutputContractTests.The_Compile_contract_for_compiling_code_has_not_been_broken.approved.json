{
  "succeeded": true,
  "base64assembly": "TVqQAAMAAAAEAAAA//8AALgAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAAA4fug4AtAnNIbgBTM0hVGhpcyBwcm9ncmFtIGNhbm5vdCBiZSBydW4gaW4gRE9TIG1vZGUuDQ0KJAAAAAAAAABQRQAATAECACHhZe8AAAAAAAAAAOAAIgALATAAAAgAAAACAAAAAAAAAiYAAAAgAAAAQAAAAABAAAAgAAAAAgAABAAAAAAAAAAEAAAAAAAAAABgAAAAAgAAAAAAAAMAQIUAABAAABAAAAAAEAAAEAAAAAAAABAAAAAAAAAAAAAAALAlAABPAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAwAAACUJQAAHAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAACAAAAAAAAAAAAAAACCAAAEgAAAAAAAAAAAAAAC50ZXh0AAAACAYAAAAgAAAACAAAAAIAAAAAAAAAAAAAAAAAACAAAGAucmVsb2MAAAwAAAAAQAAAAAIAAAAKAAAAAAAAAAAAAAAAAABAAABCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADkJQAAAAAAAEgAAAACAAUAgCAAABQFAAABAAAAAQAABgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADYAKAMAAAYoCwAACgAqIgIoDAAACgAqABMwAQALAAAAAQAAEQByAQAAcAorAAYqAEJTSkIBAAEAAAAAAAwAAAB2NC4wLjMwMzE5AAAAAAUAbAAAALABAAAjfgAAHAIAAAQCAAAjU3RyaW5ncwAAAAAgBAAAHAAAACNVUwA8BAAAEAAAACNHVUlEAAAATAQAAMgAAAAjQmxvYgAAAAAAAAACAAABRxQCAAkAAAAA+gEzABYAAAEAAAANAAAAAwAAAAMAAAAMAAAACgAAAAEAAAABAAAAAgAAAAAAgAEBAAAAAAAGAPYAywEGAEgBywEGAFAAuAEPAOsBAAAGAHsAZgEGAC8BoAEGANcAoAEGAJQAoAEGALEAoAEGABYBoAEGAGQAoAEGAPoBlAEKABgAlAEAAAAAAQAAAAAAAQABAAEAEACMAS8AMQABAAEAgQEQACgALwAxAAEAAwBQIAAAAACWAJsBJwABAF4gAAAAAIYYsgEGAAEAaCAAAAAAlgAKACsAAQAJALIBAQARALIBBgAZALIBCgApALIBEAAxALIBEAA5ALIBEABBALIBEABJALIBEABRALIBEABZALIBEABpAEYAFQBhALIBBgAuAAsALwAuABMAOAAuABsAVwAuACMAYAAuACsAlgAuADMAowAuADsArgAuAEMAuwAuAEsAlgAuAFMAlgAaAASAAAABAAAAAAAAAAAAAAAAACAAAAAEAAIAAQAAAAAAAAAeADcAAAAAAAQAAQABAAAAAAAAAB4AEQAAAAAAAAAAAAA8TW9kdWxlPgBNZXRob2QAU3lzdGVtLkNvbnNvbGUAY29uc29sZQBTYW1wbGUARXhhbXBsZQBTeXN0ZW0uUnVudGltZQBXcml0ZUxpbmUARGVidWdnYWJsZUF0dHJpYnV0ZQBBc3NlbWJseVRpdGxlQXR0cmlidXRlAFRhcmdldEZyYW1ld29ya0F0dHJpYnV0ZQBBc3NlbWJseUZpbGVWZXJzaW9uQXR0cmlidXRlAEFzc2VtYmx5SW5mb3JtYXRpb25hbFZlcnNpb25BdHRyaWJ1dGUAQXNzZW1ibHlDb25maWd1cmF0aW9uQXR0cmlidXRlAENvbXBpbGF0aW9uUmVsYXhhdGlvbnNBdHRyaWJ1dGUAQXNzZW1ibHlQcm9kdWN0QXR0cmlidXRlAEFzc2VtYmx5Q29tcGFueUF0dHJpYnV0ZQBSdW50aW1lQ29tcGF0aWJpbGl0eUF0dHJpYnV0ZQBTeXN0ZW0uUnVudGltZS5WZXJzaW9uaW5nAGNvbnNvbGUuZGxsAFByb2dyYW0AU3lzdGVtAE1haW4AU3lzdGVtLlJlZmxlY3Rpb24ALmN0b3IAU3lzdGVtLkRpYWdub3N0aWNzAFN5c3RlbS5SdW50aW1lLkNvbXBpbGVyU2VydmljZXMARGVidWdnaW5nTW9kZXMAT2JqZWN0AAAAAAAZSABlAGwAbABvACAAdwBvAHIAbABkACEAAADupK7iGSNyQq/C6WLZ97a7AAQgAQEIAyAAAQUgAQEREQQgAQEOBAABARwDBwEcCLA/X38R1Qo6AwAAAQMAABwIAQAIAAAAAAAeAQABAFQCFldyYXBOb25FeGNlcHRpb25UaHJvd3MBCAEABwEAAAAANQEAGC5ORVRDb3JlQXBwLFZlcnNpb249djIuMQEAVA4URnJhbWV3b3JrRGlzcGxheU5hbWUADAEAB2NvbnNvbGUAAAoBAAVEZWJ1ZwAADAEABzEuMC4wLjAAAAoBAAUxLjAuMAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAA2CUAAAAAAAAAAAAA8iUAAAAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOQlAAAAAAAAAAAAAAAAX0NvckV4ZU1haW4AbXNjb3JlZS5kbGwAAAAAAP8lACBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAAMAAAABDYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
  "diagnostics": []
}